tests/vcs/test_clone.py

```diff
@@ -239,5 +147,85 @@ def test_clone_should_rstrip_trailing_slash_in_repo_url(mocker, clone_dir):
     )
 
     mock_subprocess.assert_called_once_with(
-        ['git', 'clone', 'https://github.com/foo/bar'], cwd=clone_dir
+        ['git', 'clone', 'https://github.com/foo/bar'],
+        cwd=clone_dir,
+        stderr=subprocess.STDOUT
     )
+
+
+@pytest.mark.parametrize('error_message', [
+    "fatal: repository 'https://github.com/hackebro/cookiedozer' not found",
+    'hg: abort: HTTP Error 404: Not Found',
+])
+def test_clone_handles_repo_typo(mocker, clone_dir, error_message):
+    """In `clone()`, repository not found errors should raise an
+    appropriate exception.
+    """
+    mocker.patch(
+        'cookiecutter.vcs.subprocess.check_output',
+        autospec=True,
+        side_effect=[subprocess.CalledProcessError(
+            -1, 'cmd', output=error_message
+        )]
+    )
+
+    repository_url = 'https://github.com/hackebro/cookiedozer'
+    with pytest.raises(exceptions.RepositoryNotFound) as err:
+        vcs.clone(
+            repository_url,
+            clone_to_dir=clone_dir,
+            no_input=True
+        )
+
+    assert str(err.value) == (
+        'The repository {} could not be found, have you made a typo?'
+    ).format(repository_url)
+
+
+@pytest.mark.parametrize('error_message', [
+    "error: pathspec 'unknown_branch' did not match any file(s) known to git.",
+    "hg: abort: unknown revision 'unknown_branch'!",
+])
+def test_clone_handles_branch_typo(mocker, clone_dir, error_message):
+    """In `clone()`, branch not found errors should raise an
+    appropriate exception.
+    """
+    mocker.patch(
+        'cookiecutter.vcs.subprocess.check_output',
+        autospec=True,
+        side_effect=[subprocess.CalledProcessError(
+            -1, 'cmd', output=error_message
+        )]
+    )
+
+    repository_url = 'https://github.com/pytest-dev/cookiecutter-pytest-plugin'
+    with pytest.raises(exceptions.RepositoryCloneFailed) as err:
+        vcs.clone(
+            repository_url,
+            clone_to_dir=clone_dir,
+            checkout='unknown_branch',
+            no_input=True
+        )
+
+    assert str(err.value) == (
+        'The unknown_branch branch of repository '
+        '{} could not found, have you made a typo?'
+    ).format(repository_url)
+
+
+def test_clone_unknown_subprocess_error(mocker, clone_dir):
+    """In `clone()`, unknown subprocess errors should be raised."""
+    mocker.patch(
+        'cookiecutter.vcs.subprocess.check_output',
+        autospec=True,
+        side_effect=[subprocess.CalledProcessError(
```

Comment:
Actually, looks like instantiating the exception before assigning it to `side_effect` works...